File: templates/Skill-Template/SkillTemplate/Skills/Skill Template/Dialogs/Sample/SampleDialog.cs
﻿using System;
using System.Collections.Specialized;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.Bot.Builder;
using Microsoft.Bot.Builder.Dialogs;
using Microsoft.Bot.Solutions.Extensions;
using Microsoft.Bot.Solutions.Skills;
using $safeprojectname$.Dialogs.Sample.Resources;
using $safeprojectname$.Dialogs.Shared;
using $safeprojectname$.ServiceClients;

namespace $safeprojectname$.Dialogs.Sample
{
    public class SampleDialog : SkillTemplateDialog
    {
        public SampleDialog(
            SkillConfigurationBase services,
            IStatePropertyAccessor<SkillConversationState> conversationStateAccessor,
            IStatePropertyAccessor<SkillUserState> userStateAccessor,
            IServiceManager serviceManager,
            IBotTelemetryClient telemetryClient)
            : base(nameof(SampleDialog), services, conversationStateAccessor, userStateAccessor, serviceManager, telemetryClient)
        {
            var sample = new WaterfallStep[]
            {
                // NOTE: Uncomment these lines to include authentication steps to this dialog
                // GetAuthToken,
                // AfterGetAuthToken,
                PromptForName,
                GreetUser,
                End,
            };

            AddDialog(new WaterfallDialog(nameof(SampleDialog), sample));
            AddDialog(new TextPrompt(DialogIds.NamePrompt));
        
            InitialDialogId = nameof(SampleDialog);
        }

        private async Task<DialogTurnResult> PromptForName(WaterfallStepContext stepContext, CancellationToken cancellationToken)
        {
            // NOTE: Uncomment the following lines to access LUIS result for this turn.
            // var state = await ConversationStateAccessor.GetAsync(stepContext.Context);
            // var intent = state.LuisResult.TopIntent().intent;
            // var entities = state.LuisResult.Entities;

            var prompt = stepContext.Context.Activity.CreateReply(SampleResponses.NamePrompt);
            return await stepContext.PromptAsync(DialogIds.NamePrompt, new PromptOptions { Prompt = prompt });
        }

        private async Task<DialogTurnResult> GreetUser(WaterfallStepContext stepContext, CancellationToken cancellationToken)
        {
            var tokens = new StringDictionary
            {
                { "Name", stepContext.Result.ToString() },
            };

            var response = stepContext.Context.Activity.CreateReply(SampleResponses.HaveNameMessage, ResponseBuilder, tokens);
            await stepContext.Context.SendActivityAsync(response);

            return await stepContext.NextAsync();
        }

        private Task<DialogTurnResult> End(WaterfallStepContext stepContext, CancellationToken cancellationToken)
        {
            return stepContext.EndDialogAsync();
        }

        private class DialogIds
        {
            public const string NamePrompt = "namePrompt";
        }
    }
}

File: templates/Skill-Template/src/typescript/generator-botbuilder-skill/generators/app/templates/customSkill/src/dialogs/sample/_sampleDialog.ts
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

import {
        ResponseManager,
        SkillConfigurationBase } from 'bot-solution';
import {
        BotTelemetryClient,
        StatePropertyAccessor } from 'botbuilder';
import {
        DialogTurnResult,
        TextPrompt,
        WaterfallDialog,
        WaterfallStepContext } from 'botbuilder-dialogs';
import { IServiceManager } from '../../serviceClients/IServiceManager';
import { SkillDialogBase } from '../shared/skillDialogBase';
import { SampleResponses } from './sampleResponses';

import { <%=skillConversationStateNameClass%> } from '../../<%=skillConversationStateNameFile%>';

import { <%=skillUserStateNameClass%> } from '../../<%=skillUserStateNameFile%>';

/**
 * this is the sampleDialog
 */
export class SampleDialog extends SkillDialogBase {

    private tokenKey: string = 'name';

    constructor(
        services : SkillConfigurationBase,
        responseManager: ResponseManager,
        conversationStateAccessor: StatePropertyAccessor<<%=skillConversationStateNameClass%>>,
        userStateAccessor: StatePropertyAccessor<<%=skillUserStateNameClass%>>,
        serviceManager: IServiceManager,
        telemetryClient: BotTelemetryClient) {
        super(
            SampleDialog.name,
            services,
            responseManager,
            conversationStateAccessor,
            userStateAccessor,
            serviceManager,
            telemetryClient
        );

        const sample: ((sc: WaterfallStepContext) => Promise<DialogTurnResult>)[] = [
            // NOTE: Uncomment these lines to include authentication steps to this dialog
            // GetAuthToken,
            // AfterGetAuthToken,
            this.promptForName.bind(this),
            this.greetUser.bind(this),
            this.end.bind(this)
        ];

        this.addDialog(new WaterfallDialog(SampleDialog.name, sample));
        this.addDialog(new TextPrompt(dialogIds.namePrompt));

        this.initialDialogId = SampleDialog.name;
    }

    private async promptForName(stepContext: WaterfallStepContext): Promise <DialogTurnResult> {
        // NOTE: Uncomment the following lines to access LUIS result for this turn.
        // let state = await conversationStateAccessor.get(stepContext.context);
        // let intent = state.luisResult.topIntent().intent;
        // let entities = state.luisResult.entities;

        // tslint:disable-next-line:no-any
        const prompt: any = this.responseManager.getResponse(SampleResponses.responseIds.namePrompt);

        return stepContext.prompt(dialogIds.namePrompt, { prompt: prompt });
    }

    private async greetUser(stepContext: WaterfallStepContext): Promise<DialogTurnResult> {

        const tokens: Map<string, string> = new Map<string, string>();
        tokens.set(this.tokenKey, <string>stepContext.result);

        // tslint:disable-next-line:no-any
        const response: any = this.responseManager.getResponse(SampleResponses.responseIds.namePrompt, tokens);
        await stepContext.context.sendActivity(response);

        return stepContext.next();
    }

    private async end(stepContext: WaterfallStepContext): Promise<DialogTurnResult> {

        return stepContext.endDialog();
    }
}

enum dialogIds {
    namePrompt = 'namePrompt'
}
